haystack/utils/type_serialization.py

```diff
@@ -91,60 +99,61 @@ def deserialize_type(type_str: str) -> Any:
         frozenset: typing.FrozenSet,
     }
 
-    def parse_generic_args(args_str):
-        args = []
-        bracket_count = 0
-        current_arg = ""
-
-        for char in args_str:
-            if char == "[":
-                bracket_count += 1
-            elif char == "]":
-                bracket_count -= 1
-
-            if char == "," and bracket_count == 0:
-                args.append(current_arg.strip())
-                current_arg = ""
-            else:
-                current_arg += char
-
-        if current_arg:
-            args.append(current_arg.strip())
-
-        return args
-
+    # Handle generics
     if "[" in type_str and type_str.endswith("]"):
-        # Handle generics
         main_type_str, generics_str = type_str.split("[", 1)
         generics_str = generics_str[:-1]
 
         main_type = deserialize_type(main_type_str)
-        generic_args = tuple(deserialize_type(arg) for arg in parse_generic_args(generics_str))
+        generic_args = [deserialize_type(arg) for arg in _parse_generic_args(generics_str)]
 
         # Reconstruct
-        if sys.version_info >= (3, 9) or repr(main_type).startswith("typing."):
-            return main_type[generic_args]
-        else:
-            return type_mapping[main_type][generic_args]  # type: ignore
+        try:
+            if sys.version_info >= (3, 9) or repr(main_type).startswith("typing."):
+                return main_type[tuple(generic_args) if len(generic_args) > 1 else generic_args[0]]
+            else:
+                return type_mapping[main_type][tuple(generic_args) if len(generic_args) > 1 else generic_args[0]]
+        except (TypeError, AttributeError) as e:
+            raise DeserializationError(f"Could not apply arguments {generic_args} to type {main_type}") from e
 
-    else:
-        # Handle non-generics
+    # Handle non-generic types
+    # First, check if there's a module prefix
+    if "." in type_str:
         parts = type_str.split(".")
-        module_name = ".".join(parts[:-1]) or "builtins"
+        module_name = ".".join(parts[:-1])
         type_name = parts[-1]
 
         module = sys.modules.get(module_name)
-        if not module:
+        if module is not None:
             try:
                 module = thread_safe_import(module_name)
```

Comment:
Maybe I am missing something, but why reimporting the module here after checking that it's already imported? Shouldn't the condition be `if module is None`?

Reply:
Good catch you're right, my brain got this mixed up.